File: packages/core-ts/src/reliance/rm-scope.ts
// SPDX-License-Identifier: Apache-2.0
// Exact scope arithmetic for the experimental RM v1 binding (browser-safe).
//
// Quantities are decimal strings (xsd:decimal). Comparisons use exact rationals:
// value × unit factor, with mg/kg = 10^-6 kg/kg, so no floating-point tolerance can
// move an acceptance boundary (handover §6.3 rules 7-8). Unknown units or malformed
// numbers are never compared.

/** Exact non-negative decimal as numerator / 10^scale. */
export interface Decimal { readonly n: bigint; readonly scale: number }

const DECIMAL = /^(0|[1-9][0-9]*)(\.[0-9]+)?$/;
/** Supported mass-fraction units as powers of ten relative to kg/kg. */
export const RM_UNIT_EXPONENTS: Readonly<Record<string, number>> = Object.freeze({ 'kg/kg': 0, 'mg/kg': -6 });

export function parseDecimal(value: unknown): Decimal | undefined {
  if (typeof value !== 'string' || !DECIMAL.test(value)) return undefined;
  const [integer, fraction = ''] = value.split('.');
  return { n: BigInt(`${integer}${fraction}`), scale: fraction.length };
}

/** Exact quantity in kg/kg as a rational numerator / 10^scale. */
export function toKgPerKg(value: unknown, unit: unknown): Decimal | undefined {
  const decimal = parseDecimal(value);
  const exponent = typeof unit === 'string' ? RM_UNIT_EXPONENTS[unit] : undefined;
  if (decimal === undefined || exponent === undefined) return undefined;
  return exponent <= 0
    ? { n: decimal.n, scale: decimal.scale - exponent }
    : { n: decimal.n * 10n ** BigInt(exponent), scale: decimal.scale };
}

export function compareDecimal(a: Decimal, b: Decimal): -1 | 0 | 1 {
  const scale = Math.max(a.scale, b.scale);
  const x = a.n * 10n ** BigInt(scale - a.scale);
  const y = b.n * 10n ** BigInt(scale - b.scale);
  return x < y ? -1 : x > y ? 1 : 0;
}

export function addDecimal(a: Decimal, b: Decimal): Decimal {
  const scale = Math.max(a.scale, b.scale);
  return { n: a.n * 10n ** BigInt(scale - a.scale) + b.n * 10n ** BigInt(scale - b.scale), scale };
}

export function formatDecimal(value: Decimal): string {
  const digits = value.n.toString().padStart(value.scale + 1, '0');
  if (value.scale === 0) return digits;
  const fraction = digits.slice(-value.scale).replace(/0+$/, '');
  return fraction.length === 0 ? digits.slice(0, -value.scale) : `${digits.slice(0, -value.scale)}.${fraction}`;
}

type ScopeRecord = Record<string, unknown>;
interface Range { from?: unknown; to?: unknown; unit?: unknown }

/** Exact interval bounds of a scope record in kg/kg, or a reason it cannot be read. */
export function recordInterval(record: ScopeRecord): { low: Decimal; high: Decimal } | string {
  const range = record.range as Range | undefined;
  if (range === undefined) return 'Scope record has no range.';
  const low = toKgPerKg(range.from, range.unit);
  const high = toKgPerKg(range.to, range.unit);
  if (low === undefined || high === undefined) return `Unsupported or malformed range ${String(range.from)}–${String(range.to)} ${String(range.unit)}.`;
  if (compareDecimal(low, high) > 0) return 'Scope record range is reversed.';
  return { low, high };
}

const list = (value: unknown): string[] => (Array.isArray(value) ? value.filter(v => typeof v === 'string') : []);
const subset = (child: string[], parent: string[]) => child.length > 0 && child.every(item => parent.includes(item));

export interface ContainmentWitness { readonly child: string; readonly parent: string }

/**
 * Bounded projection: every child record must fit inside ONE complete parent record
 * (no splicing across records, no implicit union of adjacent records). Missing or
 * empty restricted dimensions never pass silently.
 */
export function containedIn(
  children: readonly ScopeRecord[], parents: readonly ScopeRecord[],
): { state: 'established' | 'contradicted' | 'not_established'; reason: string; witnesses: ContainmentWitness[] } {
  if (children.length === 0) return { state: 'not_established', reason: 'The projected scope has no records.', witnesses: [] };
  const witnesses: ContainmentWitness[] = [];
  for (const child of children) {
    const childRange = recordInterval(child);
    if (typeof childRange === 'string') return { state: 'not_established', reason: `${String(child.id)}: ${childRange}`, witnesses };
    const properties = list(child.allowedPropertyIris), methods = list(child.allowedMethodIris);
    if (properties.length === 0 || methods.length === 0 || typeof child.matrixIri !== 'string' ||
        typeof child.formIri !== 'string' || typeof child.quantityKindIri !== 'string') {
      return { state: 'not_established', reason: `${String(child.id)}: a restricted dimension is missing or empty.`, witnesses };
    }
    const parent = parents.find(p => {
      const parentRange = recordInterval(p);
      return typeof parentRange !== 'string'
        && p.matrixIri === child.matrixIri && p.formIri === child.formIri && p.quantityKindIri === child.quantityKindIri
        && subset(properties, list(p.allowedPropertyIris)) && subset(methods, list(p.allowedMethodIris))
        && compareDecimal(childRange.low, parentRange.low) >= 0 && compareDecimal(childRange.high, parentRange.high) <= 0;
    });
    if (parent === undefined) {
      return { state: 'contradicted', reason: `${String(child.id)} is not contained in any single parent record.`, witnesses };
    }
    witnesses.push({ child: String(child.id), parent: String(parent.id) });
  }
  return { state: 'established', reason: `Each projected record lies within one parent record (${witnesses.map(w => `${w.child} ⊆ ${w.parent}`).join('; ')}).`, witnesses };
}

File: packages/core-py/qi_vc_core/reliance/rm_scope.py
# SPDX-License-Identifier: Apache-2.0
"""Exact scope arithmetic for the experimental RM v1 binding (mirrors rm-scope.ts).

Quantities are decimal strings compared as exact rationals in kg/kg (mg/kg = 10^-6),
so no floating-point tolerance can move an acceptance boundary. Unknown units or
malformed numbers are never compared.
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from fractions import Fraction
from typing import Any, Literal

_DECIMAL = re.compile(r"^(0|[1-9][0-9]*)(\.[0-9]+)?$")
RM_UNIT_EXPONENTS: dict[str, int] = {"kg/kg": 0, "mg/kg": -6}


def parse_decimal(value: Any) -> Fraction | None:
    if not isinstance(value, str) or not _DECIMAL.fullmatch(value):
        return None
    return Fraction(value)


def to_kg_per_kg(value: Any, unit: Any) -> Fraction | None:
    decimal = parse_decimal(value)
    exponent = RM_UNIT_EXPONENTS.get(unit) if isinstance(unit, str) else None
    if decimal is None or exponent is None:
        return None
    return decimal * Fraction(10) ** exponent


def format_decimal(value: Fraction) -> str:
    """Exact decimal string for a terminating fraction (denominator 2^a 5^b)."""
    if value.denominator == 1:
        return str(value.numerator)
    scale = 0
    scaled = value
    while scaled.denominator != 1:
        scaled *= 10
        scale += 1
    digits = str(scaled.numerator).rjust(scale + 1, "0")
    fraction = digits[-scale:].rstrip("0")
    return digits[:-scale] + ("." + fraction if fraction else "")


def record_interval(record: dict[str, Any]) -> tuple[Fraction, Fraction] | str:
    scope_range = record.get("range")
    if not isinstance(scope_range, dict):
        return "Scope record has no range."
    low = to_kg_per_kg(scope_range.get("from"), scope_range.get("unit"))
    high = to_kg_per_kg(scope_range.get("to"), scope_range.get("unit"))
    if low is None or high is None:
        return (
            f"Unsupported or malformed range {scope_range.get('from')}–"
            f"{scope_range.get('to')} {scope_range.get('unit')}."
        )
    if low > high:
        return "Scope record range is reversed."
    return low, high


def _strings(value: Any) -> list[str]:
    return [v for v in value if isinstance(v, str)] if isinstance(value, list) else []


def _subset(child: list[str], parent: list[str]) -> bool:
    return bool(child) and all(item in parent for item in child)


@dataclass(frozen=True)
class Containment:
    state: Literal["established", "contradicted", "not_established"]
    reason: str
    witnesses: tuple[tuple[str, str], ...]


def contained_in(
    children: list[dict[str, Any]], parents: list[dict[str, Any]]
) -> Containment:
    """Every child record must fit inside ONE complete parent record (no splicing)."""
    if not children:
        return Containment("not_established", "The projected scope has no records.", ())
    witnesses: list[tuple[str, str]] = []
    for child in children:
        child_range = record_interval(child)
        if isinstance(child_range, str):
            return Containment(
                "not_established", f"{child.get('id')}: {child_range}", tuple(witnesses)
            )
        properties = _strings(child.get("allowedPropertyIris"))
        methods = _strings(child.get("allowedMethodIris"))
        if (
            not properties
            or not methods
            or not all(
                isinstance(child.get(k), str)
                for k in ("matrixIri", "formIri", "quantityKindIri")
            )
        ):
            return Containment(
                "not_established",
                f"{child.get('id')}: a restricted dimension is missing or empty.",
                tuple(witnesses),
            )
        match = None
        for parent in parents:
            parent_range = record_interval(parent)
            if (
                not isinstance(parent_range, str)
                and parent.get("matrixIri") == child.get("matrixIri")
                and parent.get("formIri") == child.get("formIri")
                and parent.get("quantityKindIri") == child.get("quantityKindIri")
                and _subset(properties, _strings(parent.get("allowedPropertyIris")))
                and _subset(methods, _strings(parent.get("allowedMethodIris")))
                and child_range[0] >= parent_range[0]
                and child_range[1] <= parent_range[1]
            ):
                match = parent
                break
        if match is None:
            return Containment(
                "contradicted",
                f"{child.get('id')} is not contained in any single parent record.",
                tuple(witnesses),
            )
        witnesses.append((str(child.get("id")), str(match.get("id"))))
    listed = "; ".join(f"{c} ⊆ {p}" for c, p in witnesses)
    return Containment(
        "established",
        f"Each projected record lies within one parent record ({listed}).",
        tuple(witnesses),
    )
